ext/nio4r/monitor.c

```diff
@@ -115,9 +129,9 @@ static VALUE NIO_Monitor_initialize(VALUE self, VALUE io, VALUE interests, VALUE
     rb_ivar_set(self, rb_intern("interests"), interests);
     rb_ivar_set(self, rb_intern("selector"), selector_obj);
 
-    Data_Get_Struct(selector_obj, struct NIO_Selector, selector);
+    selector = NIO_Selector_unwrap(selector_obj);
 
-    monitor->self = self;
+    RB_OBJ_WRITE(self, &monitor->self, self);
```

Comment:
Just for my own edification, what is the nature of `RB_OBJ_WRITE` and when should it be used?

Reply:
It's a helper to assign a reference and trigger write barriers, it is required to allow `Monitor` objects to be marked as `WB_PROTECTED` which is what allow them to reach the old generation.

In short, when an object get to the old generation, it's no longer marked during minor GC. So if you create a reference from an old to a young object, the GC won't see it and mayincorrectly assume the young object is no longer referenced by anyone.

Write barriers are a way to tell the GC: "Hey, this old object created a new reference, you should check it".

In this specific instance, it's actually not really required because it's an object creating a reference to itself, so from a GC point of view, it's a bit useless. Similarly when you assign to an immediate like `Qnil`, it's not a big deal if you forget it.

But I like to always do it as I think consistency here help avoiding future patches to forget it.